ssl/quic/quic_tls_api.c

```diff
@@ -0,0 +1,187 @@
+            if (qtcb->crypto_send_cb == NULL)
+                qtcb->crypto_send_cb = OSSL_FUNC_SSL_QUIC_TLS_crypto_send(qtdis);
+            break;
+        case OSSL_FUNC_SSL_QUIC_TLS_CRYPTO_RECV_RCD:
+            if (qtcb->crypto_recv_rcd_cb == NULL)
+                qtcb->crypto_recv_rcd_cb =
+                    OSSL_FUNC_SSL_QUIC_TLS_crypto_recv_rcd(qtdis);
+            break;
+        case OSSL_FUNC_SSL_QUIC_TLS_CRYPTO_RELEASE_RCD:
+            if (qtcb->crypto_release_rcd_cb == NULL)
+                qtcb->crypto_release_rcd_cb =
+                    OSSL_FUNC_SSL_QUIC_TLS_crypto_release_rcd(qtdis);
+            break;
+        case OSSL_FUNC_SSL_QUIC_TLS_YIELD_SECRET:
+            if (qtcb->yield_secret_cb == NULL)
+                qtcb->yield_secret_cb =
+                    OSSL_FUNC_SSL_QUIC_TLS_yield_secret(qtdis);
+            break;
+        case OSSL_FUNC_SSL_QUIC_TLS_GOT_TRANSPORT_PARAMS:
+            if (qtcb->got_transport_params_cb == NULL)
+                qtcb->got_transport_params_cb =
+                    OSSL_FUNC_SSL_QUIC_TLS_got_transport_params(qtdis);
+            break;
+        case OSSL_FUNC_SSL_QUIC_TLS_ALERT:
+            if (qtcb->alert_cb == NULL)
+                qtcb->alert_cb =
+                    OSSL_FUNC_SSL_QUIC_TLS_alert(qtdis);
+            break;
+        }
+    }
+
+    if (qtcb->crypto_send_cb == NULL
+            || qtcb->crypto_recv_rcd_cb == NULL
+            || qtcb->crypto_release_rcd_cb == NULL
+            || qtcb->yield_secret_cb == NULL
+            || qtcb->got_transport_params_cb == NULL
+            || qtcb->alert_cb == NULL) {
+        ERR_raise(ERR_LIB_SSL, SSL_R_MISSING_QUIC_TLS_FUNCTIONS);
+        return 0;
+    }
+
+    return 1;
+}
+
+int SSL_set_quic_tls_cbs(SSL *s, const OSSL_DISPATCH *qtdis, void *arg)
+{
+    SSL_CONNECTION *sc = SSL_CONNECTION_FROM_SSL(s);
+    QUIC_TLS_ARGS qtlsargs;
+
+    if (!SSL_is_tls(s)) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
+        return 0;
+    }
+
+    if (!tls_callbacks_from_dispatch(&sc->qtcb, qtdis))
+        /* ERR_raise already called*/
+        return 0;
+
+    sc->qtarg = arg;
+
+    ossl_quic_tls_free(sc->qtls);
+    qtlsargs.s = s;
+    qtlsargs.crypto_send_cb = crypto_send_cb;
+    qtlsargs.crypto_send_cb_arg = s;
+    qtlsargs.crypto_recv_rcd_cb = crypto_recv_rcd_cb;
+    qtlsargs.crypto_recv_rcd_cb_arg = s;
+    qtlsargs.crypto_release_rcd_cb = crypto_release_rcd_cb;
+    qtlsargs.crypto_release_rcd_cb_arg = s;
+    qtlsargs.yield_secret_cb = yield_secret_cb;
+    qtlsargs.yield_secret_cb_arg = s;
+    qtlsargs.got_transport_params_cb = got_transport_params_cb;
+    qtlsargs.got_transport_params_cb_arg = s;
+    qtlsargs.handshake_complete_cb = NULL;
+    qtlsargs.handshake_complete_cb_arg = NULL;
+    qtlsargs.alert_cb = alert_cb;
+    qtlsargs.alert_cb_arg = s;
+    qtlsargs.is_server = sc->server;
+    qtlsargs.ossl_quic = 0;
+    sc->qtls = ossl_quic_tls_new(&qtlsargs);
```

Comment:
Does there need to be an extra call to ossl_quic_tls_free in ossl_quic_free here to avoid the memory leak reported in CI?

Reply:
It needs to be in ssl3_free. Already fixed in the latest fixup I just pushed a few moments ago.